firestore.rules

```diff
@@ -0,0 +1,91 @@
+// NOTE: to deploy only these rules run
+// `npm run deploy:firestore:rules`
+
+rules_version = '2';
+service cloud.firestore {
+  match /databases/{database}/documents {
+
+    function isAuthed() {
+      return request.auth != null;
+    }
+
+    // function hasClass() {
+    //   return isAuthed() && request.auth.token.class_hash != null;
+    // }
+
+    // function hasClassHash(classHash) {
+    //   return request.auth.token.class_hash == classHash;
+    // }
+
+    function hasRole(role) {
+      return isAuthed() && request.auth.token.user_type == role;
+    }
+
+    function hasUserId() {
+      return isAuthed() && request.auth.token.platform_user_id != null;
+    }
+
+    // function matchUserId(userId) {
+    //   return hasUserId() && string(request.auth.token.platform_user_id) == userId;
+    // }
+
+    function matchFirebaseUserId(userId) {
+      return isAuthed() && request.auth.uid == userId;
+    }
+
+    function isAuthedTeacher() {
+      return hasUserId() && hasRole("teacher");
+    }
+
+    function docMatchUserId() {
+      return isAuthed() &&
+        string(request.auth.token.platform_user_id) == resource.data.uid;
+    }
+
+    function classInClasses() {
+      return isAuthed() && request.auth.token.class_hash in resource.data.classes;
+    }
+
+    //
+    // portal-authenticated secure access rules
+    //
+    match /authed/{portal} {
+      allow read, write: if isAuthedTeacher();
+
+        match /mcsupports/{docId} {
+          // any portal-authenticated teacher can create supports
+          allow create: if isAuthedTeacher();
```

Comment:
One thing you can do here to lock it down more is to restrict users from creating supports with some other teacher's `data.uid`.  To check that you look at `request.resource.data.uid`.  The `request.resource` is the document being sent up in the request.

I think this same approach can be applied to the update method, so a teacher can't change the uid to point to some other teacher.

I don't think this form of locking down can be used with the class_hash since it is list and we don't know the full list of class_hashes of the current user. And in theory the teacher could be removing the support from the current class they are auth'd with, but they still want the support available to their other classes.

Reply:
done